### ISSUE TITLE
Focusout event not triggered when test window is not focused

### ISSUE THREAD

**Issue by schorke** [emberjs/ember-test-helpers#692]
So, I have a simple Ember component input which has the built-in `focusOut` Ember event. In an integration test, using `ember-test-helpers` `triggerEvent(input, 'blur')` on this input or `fillIn(other-input, 'blbl')` in another input does not trigger the `focusOut` event if the window test is not focused while running.

I created a new Ember project to demonstrate this bug (see `tests/integration/my-component-test.js`), you can check it out here => [test-focusout](https://github.com/schorke/test-focusout)

### How to test
1. `ember t -s --filter='Integration | Component | my-component'`
2. Open dev tools in test window.
3. Click in console and refresh page (click inside test window if you want to re-focusin).
4. Same behaviour when running the test while focusing another window.
Reproduced on Chrome 76.0.3809.87 and Firefox 68.0.1

Is that the expected behaviour ?

**Comment by rwjblue** [comment]
> Is that the expected behaviour ?

Sadly, I believe it is. Specifically, we can't always trigger focus events when the window itself doesn't have focus.  The best way to confirm this would be to make a simpler reproduction with plain HTML / JS (no Ember, no jQuery, etc) and see if we can dispatch the events when the window isn't focused. If we **can** then we can likely make it work here too.

**Comment by trek** [comment]
Not that it's a fix, but we worked around this by adding a `testWithBrowserFocus` wrapper for `test` that we use in these cases.

Here we preferred false positives.

```
import { test } from 'qunit';
/*
  Tests that need to verify HTMLElement.focus() behavior will
  fail when run in a window that isn't focused, but pass correctly
  when the window is focused or running headless (e.g. in CI).

  This helper checks if the window is focused. If it's not,
  the test will pass.
*/
export function testWithBrowserFocus(description, callback) {
  test(description, async function(assert) {
    if (!document.hasFocus()) {
      assert.ok(true);
      return;
    }
    await callback.call(this, assert);
  });
}
```

And it works just like `test`:

```
  testWithBrowserFocus('opening a chat marks it as read', async function(assert) {});

```

**Comment by ohcibi** [comment]
I noticed that the current `fillIn` implementation does in fact fire `focus` and `focusin` but no `focusout` event. Is there a reason why the first two are fired but not the out event?

**Comment by rwjblue** [comment]
@ohcibi - I thats a different issue than the one reported here, would you mind reporting that separately? I think we should fix...

**Comment by rwjblue** [comment]
I'm not sure what we can do other than @trek's suggestion. For the most part this is out of our control...

**Comment by ohcibi** [comment]
@rwjblue No it is no other issue. I rather found the cause of this issue, lemme phrase it once again: **`focusout` is never called when using `fillIn`. No matter if the browser window has focus or not.** The only function of this plugin that fires a `focusout` event is `blur` (found by `grep -iR focusout *` or githubs repository search) but no other function (especially not `fillIn`) is calling that, so it is impossible for this event to be fired by using `fillIn`. I would suggest to extend this issue right here instead of creating a new one. The way it looks to me is that it is simply another issue with the docs and certainly no reason to "fix" bugs that are none.

**Comment by rwjblue** [comment]
@ohcibi - Ya, you are right sorry about that. I'm not sure what the best path forward here is though. `fillIn` / `typeIn` could theoretically focus out at the end, but I'm not 100% sure that is the correct behavior for us. 🤔 

@ro0gr / @Turbo87 - Thoughts?

**Comment by Turbo87** [comment]
hmm, I don't think we should blur at the end. what we could potentially do is blur at the start if the focus is on another element. or am I missing something?

**Comment by rwjblue** [comment]
You are suggesting that we would effectively to do:

```js
if (document.activeElement) { 
  await blur(document.activeElement);
}
```

Inside `fillIn` / `typeIn` / `click`?

**Comment by ro0gr** [comment]
I agree with @Turbo87. Focusing out after `fillIn`(and especially `typeIn`) may be un-desirable, for example in case if I want to test the field right after typing in it. And focus out the active element(if it's not the target) before `fillIn` / `typeIn` / `click` seems reasonable to me.

**Comment by Turbo87** [comment]
```
if (document.activeElement !== targetElement) { 
  await blur(document.activeElement);
}
```

something like that, yeah

**Comment by ro0gr** [comment]
running ahead the horses: it should probably be a part of `__focus__`, as the very first step.

**Comment by rwjblue** [comment]
Ya, that makes sense to me.

**Comment by rwjblue** [comment]
Lets do it!

**Comment by burritoIand** [comment]
just hit this while trying to test that an input keeps focus after `await fillIn()`. Currently it loses focus on test only

**Comment by rwjblue** [comment]
Ya, definitely still would love to make progress here.
